### ISSUE TITLE
Allowing disabling percent encoding for some HTTP header fields

### ISSUE THREAD

**Issue by zunda-pixel** [apple/swift-openapi-generator#529]
### Question

I want to redirect to `domain2.com` from `domain1.com`.

I can redirect same domain.
ex: from`domain1.com/login` to `domain1.com/account`.

```yaml
responses:
  '303':
    headers:
      location:
        schema:
          type: string
```

```swift
func login(_ input: Operations.login.Input) async throws -> Operations.login.Output {
  // login
  return .seeOther(.init(headers: .init(location: "https:/example2.com")))
}

```

**Comment by czechboy0** [comment]
Hi @zunda-pixel,

just to confirm, you're writing a server here, right? You can return a completely arbitrary URL in the response's Location header, can you clarify why you say you can't redirect to a different domain?

**Comment by zunda-pixel** [comment]
Yes, I am writing. server.

Base url is keeping...

`http://127.0.0.1:8080/login` -> `http://127.0.0.1:8080/https%3A%2F%2Fexample2.com`

**Comment by czechboy0** [comment]
Hmm, `baseURL` is only a client-side concept in Swift OpenAPI Generator, not a server one. Can you describe in detail how you're testing this and how you're getting this URL?

**Comment by zunda-pixel** [comment]
This is sample repository that has issue I told. Please check.
https://github.com/zunda-pixel/LoginServer

**Comment by czechboy0** [comment]
Can you clarify what the issue is? The code in the project all looks correct.

What are the steps you're taking, what is the result you see, and what is the result you expect? That'll help us understand where the mismatch is.

**Comment by zunda-pixel** [comment]
## Current  Result

1. http://localhost:8080/login
2. http://127.0.0.1:8080/https%3A%2F%2Fapple.com

## Expecting Result

1. http://localhost:8080/login
2. https://apple.com

**Comment by czechboy0** [comment]
Which HTTP client are you using? A web browser? curl?

**Comment by zunda-pixel** [comment]
I use a web browser.

**Comment by czechboy0** [comment]
Please try this with just pure Vapor, this might be a behavior caused by the web browser (redirecting from HTTP to a HTTPS URL).

Try with just curl to verify that the Location header is being received correctly:

```
curl -v http://localhost:8080/login
```

**Comment by czechboy0** [comment]
Thank you @zunda-pixel, I was able to isolate the issue.

The problem is that OpenAPI-defined headers are serialized according to the rules of RFC6570 (details [here](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#style-values)), which dictate that non-reserved characters need to be percent encoded. However, in the `Location` header, that causes the URL in the response header to be percent-encoded, and the web browser client doesn't remove the percent encoding, it seems.

As a workaround, add a middleware that removes the percent encoding of the Location header:

```swift
import OpenAPIRuntime
import OpenAPIVapor
import Vapor
import HTTPTypes

struct Handler: APIProtocol {
  func login(_ input: Operations.login.Input) async throws -> Operations.login.Output {
    return .seeOther(.init(headers: .init(location: "https://apple.com")))
  }
}

@main struct LoginServer {
  static func main() async throws {
    let app = Vapor.Application()
    let transport = VaporTransport(routesBuilder: app)
    let handler = Handler()
    try handler.registerHandlers(on: transport, middlewares: [
        UnescapeLocationHeaderMiddleware()
    ])
    try await app.execute()
  }
}

struct UnescapeLocationHeaderMiddleware: ServerMiddleware {
    func intercept(
        _ request: HTTPRequest,
        body: HTTPBody?,
        metadata: ServerRequestMetadata,
        operationID: String,
        next: (HTTPRequest, HTTPBody?, ServerRequestMetadata) async throws -> (HTTPResponse, HTTPBody?)
    ) async throws -> (HTTPResponse, HTTPBody?) {
        var (response, responseBody) = try await next(request, body, metadata)
        guard let location = response.headerFields[.location] else {
            return (response, responseBody)
        }
        response.headerFields[.location] = location.removingPercentEncoding
        return (response, responseBody)
    }
}
```

Now, thanks for reporting this. It's a bit troubling, and I suspect we'll need some way to make this more compatible with clients that don't percent-decode header fields.

If you don't mind, I'll repurpose this issue to track improving Swift OpenAPI Generator this way and rename it.

**Comment by zunda-pixel** [comment]
Thank you for looking into this issue, and providing a workaround.

**Comment by liambutler-lawrence** [comment]
👋 Is there any progress on this issue? We just started using `swift-openapi-generator` to generate a Swift client for our existing Lambda-based backend, and were surprised that Swift decided to unilaterally URL-encode header params, which is not part of the OpenAPI [spec](https://swagger.io/docs/specification/v3_0/describing-parameters/#header-parameters).

**Comment by czechboy0** [comment]
Hi @liambutler-lawrence,

percent-encoding of special characters in header fields is prescribed by OpenAPI, see this comment for details and links to the RFC: https://github.com/apple/swift-openapi-generator/issues/529#issuecomment-1952485745

This issue tracks some possible ways to opt out of the behavior, because some services write their OpenAPI docs after their initial implementation, and they don't correctly remove percent encoding. That's a bug on any service that does that, but it's common enough that we're open to helping our users handle that situation better to interop with such legacy services.

If you'd like to see progress here, please propose a solution that we can discuss (how exactly should the opt out work?), and maybe you could ever open a PR for it later 🙂

**Comment by liambutler-lawrence** [comment]
@czechboy0 Thanks for following up and linking that doc. I stand corrected; however, we currently have no plans to change our server our other clients to (unnecessarily IMHO) URL-encode header params. I would suggest adding a configuration flag that can be passed when initializing the generated Client, e.g.

```swift
let client = Client(serverURL: foo, configuration: .init(urlEncodeHeaderParams: false), transport: bar)
```

**Comment by czechboy0** [comment]
Our goal is to conform to the OpenAPI specification. For non-compliant implementations, there's the middleware solution from above. If we support some sort of opt-out, it's likely that it'd be specific to certain headers that historically haven't used percent encoding. But for any new headers, a compliant OpenAPI implementation should include percent encoding. Unilaterally diverging from the specification would break our compatibility with other OpenAPI tools.

**Comment by Rool** [comment]
I've been reading the specs on that [encoding](https://swagger.io/specification/#appendix-d-serializing-headers-and-cookies), 

> [RFC6570](https://www.rfc-editor.org/rfc/rfc6570)'s percent-encoding behavior is not always appropriate for in: "header" and in: "cookie" parameters. In many cases, it is more appropriate to use content with a media type such as text/plain and require the application to assemble the correct string.

If I try something like
```
"parameters": [
    {
      "name": "If-None-Match",
      "in": "header",
      "description": "ETag value to check if the resource has changed",
      "required": false,
      "content": {
        "text/plain": {
          "schema": {
            "type": "string"
          }
        }
      }
    }
  ]
```
the generator does not generate an initializer for this header input.

**Comment by czechboy0** [comment]
Hi @Rool, does the generator produce a diagnostic explaining why it skipped this parameter? It should.

**Comment by Rool** [comment]
Input:
```
  "paths": {
    "/pft.json": {
      "get": {
        "summary": "Retrieve Patient Friendly Terms",
        "description": "Returns a dictionary of medical concepts keyed by unique identifiers.",
        "operationId": "getPft",
        "parameters": [
          {
            "name": "If-None-Match",
            "in": "header",
            "description": "ETag value to check if the resource has changed",
            "required": false,
            "content": {
              "text/plain": {
                "schema": {
                  "type": "string"
                }
              }
            }
          }
        ],
....
```
Logs:
```
Feature "Unsupported content" is not supported, skipping [context: foundIn=Operations.getPft.Input.Headers (#/paths/pft.json/GET/header)]

```

**Comment by Rool** [comment]
If I change the `text/plain` to `application/json`, I do see the generated If-None-Matched headers, but they are percent encoded when transmitted. 
```
/// ETag value to check if the resource has changed
                ///
                /// - Remark: Generated from `#/paths/pft.json/GET/header/If-None-Match`.
                public var If_hyphen_None_hyphen_Match: Swift.String?
                public var accept: [OpenAPIRuntime.AcceptHeaderContentType<Operations.getPft.AcceptableContentType>]
```

**Comment by czechboy0** [comment]
I see - so Swift OpenAPI Generator doesn't currently support using the custom content + plain text combination as a way to get around percent encoding.

For now, I'd recommend the middleware workaround outlined here: https://github.com/apple/swift-openapi-generator/issues/529#issuecomment-1952485745

But longer term, I'd like to see us support the content + plain text solution, especially since it's called out in the Swagger docs (even if it's not part of the official specification).

**Comment by romnes** [comment]
Looks like the [OpenAPI Spec (3.1.2)](https://spec.openapis.org/oas/v3.1.2.html) specifically calls out special casing for `header` and `cookie` now:

> When serializing in: "header" parameters with schema, URI percent-encoding MUST NOT be applied; if using an RFC6570 implementation that automatically applies it, it MUST be removed before use. Implementations MUST pass header values through unchanged rather than attempting to automatically quote header values, as the quoting rules vary too widely among different headers; see [Appendix D](https://spec.openapis.org/oas/v3.1.2.html#appendix-d-serializing-headers-and-cookies) for guidance on quoting and escaping.
> 
> Serializing with schema is NOT RECOMMENDED for in: "cookie" parameters; see [Appendix D](https://spec.openapis.org/oas/v3.1.2.html#appendix-d-serializing-headers-and-cookies) for details.

**Comment by czechboy0** [comment]
Yup, with OpenAPI 3.1.2 and 3.2.0 we got the clarification we needed. However, making this change will need to be done carefully - we'll need to consider if we just switch the behavior under everyone, or if this will need to be opted in somehow.
